### ISSUE TITLE
argon2: std feature pulls in rand number generator

### ISSUE THREAD

**Issue by webmaster128** [RustCrypto/password-hashes#250]
I'm not sure to what degree this is expected, but if I enable the `std` feature in argon2, it pulls in a random number generator via `getrandom`. This is unfortunately not available in all std environments, especially when compiling to Wasm.

```
argon2 = { version = "0.3.1", default-features = false, features = ["std"] }
```

causes a compile fail with target wasm32-unknown-unknown because

```
$ cargo tree -i getrandom --target wasm32-unknown-unknown
getrandom v0.2.3
└── rand_core v0.6.3
    └── password-hash v0.3.2
        └── argon2 v0.3.1
            └── argon2id-js v0.1.0 (/projects/cosmjs/wasm/argon2id-js)
```

Seems like `password-hash` gets pulled in implicitly with default arguments or something like that.

Adding `resolver = "2"` does not seems to change things.

Using alloc instead lets me compile.

**Comment by tarcieri** [comment]
It's definitely expected. For any of our crates that rely on an RNG, we're trying to make it possible to pull everything required in without a complicated recipe that involves things like upgrading `rand_core` in lockstep. Our experience has been that's confusing to end users.

It's unfortunate there's no unified RNG option for WASM, however. It seems like the best we can offer is explicitly linking to `getrandom` and toggling on something like the `js` feature transitively.

I suppose we should split out something like a `rand` feature from the `std` feature that toggles on `rand_core/std` at the very least.

**Comment by webmaster128** [comment]
> I suppose we should split out something like a rand feature from the std feature that toggles on rand_core/std at the very least.

What is important for me here is to keep the rng out. Adding the rnd implicitly via `std` surprised me, especially because there are Argon2 features "password-hash" and "rand" already that I did not enable.

**Comment by tarcieri** [comment]
I think the best way forward here is to properly split out `rand` and `rand_core` features, and have `rand` active `rand_core/std`

**Comment by webmaster128** [comment]
> Also I would argue that instead of relying on user-written JS code to pass entropy to WASM you should prefer using `getrandom` either way (unless you bother with wrapping WASM modules manually without relying on `wasm-bindgen`).

TL;DR: It's complicated.

That sounds so simple in theory. Unfortunately wasm-bindgen is an best effort approch in a horrible mess of JS module systems with bad support for Wasm asset compilation. Last time I tried it was impossible to write a JS lib that includes Wasm and works with webpack4+5 and CommonJS. If you try to achive async Wasm compiling it gets even worse. There is still no _embed Wasm blob as text_ target, which is the only compatible solution (used by e.g. long.js and libsodium.js). If you need to support multiple JS environments, the best you get out of wasm-bindgen is an interface on the Wasm side and a draft of a JS wrapper that you then need to modify. Now with getrandom+js you suddenly get not one interface but two interfaces to worry about (call into Wasm and call out of Wasm).

I'm not saying every application developer should mess around with entropy sources. But a well maintained JS library that bundles Wasm is certainly able to use JS system RNGs and pass the random data to Wasm from outside.

**Comment by tarcieri** [comment]
It seems like what we should actually do here is have a `getrandom` feature which forwards to `rand_core/getrandom`.

**Comment by webmaster128** [comment]
> which activates `rand_core/getrandom`.

I think the problem is `rand_core/std` [already pull in getrandom](https://github.com/rust-random/rand/blob/master/rand_core/Cargo.toml#L27). So we either need to convice the authors this is not a good idea or deactivate `rand_core/std` when the `getrandom` feature is not set.

**Comment by tarcieri** [comment]
As you said, this issue is ultimately that `rand_core/std` activates `getrandom`. We can't do anything about it here.

Given that, closing this issue.
